File: webapp/lib/api/calls.ts
import { z } from "zod";

import type { CallDetail, CallListResponse, CallSummary } from "@/lib/dto";

const CallSummaryApiSchema = z.object({
  id: z.string(),
  assistant_id: z.string(),
  customer_number: z.string().nullish(),
  status: z.string(),
  started_at: z.string().nullish(),
  ended_at: z.string().nullish(),
  duration_seconds: z.number().nullish(),
  cost: z.number().nullish(),
  transcript_preview: z.string().nullish(),
  sentiment: z.number().nullish().optional(),
});

const CallDetailApiSchema = CallSummaryApiSchema.extend({
  transcript: z.string().nullish(),
  metadata: z.record(z.any()).nullish(),
  recording_url: z.string().nullish(),
});

type CallSummaryApi = z.infer<typeof CallSummaryApiSchema>;
type CallDetailApi = z.infer<typeof CallDetailApiSchema>;

function mapCallSummary(payload: CallSummaryApi): CallSummary {
  return {
    id: payload.id,
    assistantId: payload.assistant_id,
    customerNumber: payload.customer_number ?? undefined,
    status: payload.status,
    startedAt: payload.started_at ?? undefined,
    endedAt: payload.ended_at ?? undefined,
    durationSeconds: payload.duration_seconds ?? undefined,
    cost: payload.cost ?? undefined,
    transcriptPreview: payload.transcript_preview ?? undefined,
    sentiment: payload.sentiment ?? undefined,
  };
}

function mapCallDetail(payload: CallDetailApi): CallDetail {
  return {
    ...mapCallSummary(payload),
    transcript: payload.transcript ?? undefined,
    metadata: payload.metadata ?? undefined,
    recordingUrl: payload.recording_url ?? undefined,
  };
}

export async function listCalls(params: { limit?: number; status?: string } = {}): Promise<CallListResponse> {
  const search = new URLSearchParams();
  if (params.limit) search.set("limit", params.limit.toString());
  if (params.status) search.set("status", params.status);

  const res = await fetch(`/api/calls${search.toString() ? `?${search.toString()}` : ""}`, { cache: "no-store" });
  if (!res.ok) throw new Error("Failed to fetch calls");

  const data = await res.json();
  const parsed = z
    .object({
      calls: z.array(CallSummaryApiSchema).default([]),
      total: z.number().optional(),
    })
    .parse(data);

  const mappedCalls = parsed.calls.map(mapCallSummary);
  return {
    calls: mappedCalls,
    total: parsed.total ?? mappedCalls.length,
  };
}

export async function getCall(callId: string): Promise<CallDetail> {
  const res = await fetch(`/api/calls/${callId}`, { cache: "no-store" });
  if (!res.ok) throw new Error("Failed to fetch call detail");
  const data = await res.json();
  const parsed = CallDetailApiSchema.parse(data);
  return mapCallDetail(parsed);
}

export async function getCallRecording(callId: string): Promise<{ recordingUrl: string }> {
  const res = await fetch(`/api/calls/${callId}/recording`, { cache: "no-store" });
  if (!res.ok) throw new Error("Recording not available");
  const data = await res.json();
  const schema = z.object({ recording_url: z.string() });
  const parsed = schema.parse(data);
  return { recordingUrl: parsed.recording_url };
}

File: api/src/presentation/api/v1/routes/calls.py
"""Calls REST endpoints."""

from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from api.src.infrastructure.database.session import get_session
from api.src.infrastructure.persistence.repositories.call_repository import (
    get_call_by_id,
    get_recent_calls,
)
from api.src.presentation.dependencies.auth import CurrentTenant, get_current_tenant

router = APIRouter(prefix="/calls", tags=["calls"])


@router.get("")
async def list_calls(
    limit: int = Query(50, ge=1, le=200),
    status: Optional[str] = Query(None),
    current: CurrentTenant = Depends(get_current_tenant),
    session: AsyncSession = Depends(get_session),
):
    """
    List recent calls with optional status filter.

    Query params:
    - limit: Max number of calls (1-200)
    - status: Filter by status (in-progress, ended, failed)
    """

    calls = await get_recent_calls(session, tenant_id=str(current.tenant.id), limit=limit)

    if status:
        calls = [call for call in calls if call.status == status]

    return {
        "calls": [
            {
                "id": call.id,
                "assistant_id": call.assistant_id,
                "customer_number": call.customer_number,
                "status": call.status,
                "started_at": call.started_at.isoformat() if call.started_at else None,
                "ended_at": call.ended_at.isoformat() if call.ended_at else None,
                "duration_seconds": call.duration_seconds,
                "cost": call.cost,
                "transcript_preview": call.transcript[:200] if call.transcript else None,
            }
            for call in calls
        ],
        "total": len(calls),
    }


@router.get("/{call_id}")
async def get_call_detail(
    call_id: str,
    current: CurrentTenant = Depends(get_current_tenant),
    session: AsyncSession = Depends(get_session),
):
    """
    Get full call details including transcript.
    """

    call = await get_call_by_id(session, call_id)
    if not call or str(call.tenant_id) != str(current.tenant.id):
        raise HTTPException(status_code=404, detail="Call not found")

    return {
        "id": call.id,
        "assistant_id": call.assistant_id,
        "customer_number": call.customer_number,
        "status": call.status,
        "started_at": call.started_at.isoformat() if call.started_at else None,
        "ended_at": call.ended_at.isoformat() if call.ended_at else None,
        "duration_seconds": call.duration_seconds,
        "cost": call.cost,
        "transcript": call.transcript,
        "metadata": call.metadata,
        "recording_url": call.metadata.get("recordingUrl") if isinstance(call.metadata, dict) else None,
    }


@router.get("/{call_id}/recording")
async def get_call_recording(
    call_id: str,
    current: CurrentTenant = Depends(get_current_tenant),
    session: AsyncSession = Depends(get_session),
):
    """
    Get recording URL for a call.
    """

    call = await get_call_by_id(session, call_id)
    if not call or str(call.tenant_id) != str(current.tenant.id):
        raise HTTPException(status_code=404, detail="Call not found")

    recording_url = call.metadata.get("recordingUrl") if isinstance(call.metadata, dict) else None
    if not recording_url:
        raise HTTPException(status_code=404, detail="Recording not available")

    return {"recording_url": recording_url}


__all__ = ["router"]
